rust/lance/src/index/vector/builder.rs

```diff
@@ -0,0 +1,296 @@
+use itertools::Itertools;
+use lance_core::{Error, Result, ROW_ID};
+use lance_file::v2::{reader::FileReader, writer::FileWriter};
+use lance_index::{
+    pb,
+    vector::{
+        graph::memory::InMemoryVectorStorage,
+        ivf::{
+            storage::{IvfData, IVF_METADATA_KEY},
+            IvfBuildParams,
+        },
+        quantizer::{Quantization, Quantizer},
+        transform::Transformer,
+        v3::{
+            shuffler::{IvfShuffleReader, IvfShuffler},
+            storage::{StorageBuilder, VectorStore},
+            subindex::IvfSubIndex,
+        },
+    },
+    INDEX_AUXILIARY_FILE_NAME, INDEX_FILE_NAME,
+};
+use lance_io::{
+    object_store::ObjectStore,
+    scheduler::ScanScheduler,
+    stream::{RecordBatchStream, RecordBatchStreamAdapter},
+    ReadBatchParams,
+};
+use lance_linalg::distance::DistanceType;
+use object_store::path::Path;
+use snafu::{location, Location};
+use tempfile::TempDir;
+
+use crate::{dataset::index, Dataset};
+
+use super::{utils, Ivf};
+
+pub struct IvfIndexBuilder<S: IvfSubIndex, Q: Quantization + Clone> {
+    // source
+    dataset: Dataset,
+    column: String,
+    distance_type: DistanceType,
+    shuffler: Box<dyn IvfShuffler>,
+    index_dir: String,
+    ivf_params: IvfBuildParams,
+    sub_index_params: S::BuildParams,
+    sub_index: S,
+    quantizer: Q,
+    temp_dir: Path,
+}
+
+impl<S: IvfSubIndex, Q: Quantization + Clone> IvfIndexBuilder<S, Q> {
+    pub fn new(
+        dataset: Dataset,
+        column: String,
+        distance_type: DistanceType,
+        shuffler: Box<dyn IvfShuffler>,
+        index_dir: String,
+        ivf_params: IvfBuildParams,
+        sub_index_params: S::BuildParams,
+        sub_index: S,
+        quantizer: Q,
+    ) -> Result<Self> {
+        let temp_dir = TempDir::new()?;
+        let temp_dir = Path::try_from(temp_dir.path().to_str().unwrap())?;
+        Ok(Self {
+            dataset,
+            column,
+            distance_type,
+            shuffler,
+            index_dir,
+            ivf_params,
+            sub_index_params,
+            sub_index,
+            quantizer,
+            temp_dir,
+        })
+    }
+
+    pub async fn build(&self) -> Result<()> {
+        // step 1. train IVF
+        let ivf = self.load_or_build_ivf().await?;
+
+        // step 2. shuffle data
+        let reader = self.shuffle_data(ivf.centroids.clone()).await?;
+        let partition_build_order = (0..self.ivf_params.num_partitions)
+            .map(|partition_id| reader.partiton_size(partition_id))
+            .collect::<Result<Vec<_>>>()?
+            // sort by partition size in descending order
+            .into_iter()
+            .enumerate()
+            .map(|(idx, x)| (x, idx))
+            .sorted()
+            .rev()
+            .map(|(_, idx)| idx)
+            .collect::<Vec<_>>();
+
+        // step 3. build sub index
+        for &partition in &partition_build_order {
```

Comment:
we probably want a way to control concurrency behavior here. For some index types it make sense to build one by one, for other index types we may want concurrent builds